lib/matplotlib/colors.py

```diff
@@ -2007,33 +2007,23 @@ def from_levels_and_colors(levels, colors, extend='neither'):
     cmap : `~matplotlib.colors.Normalize`
     norm : `~matplotlib.colors.Colormap`
     """
-    colors_i0 = 0
-    colors_i1 = None
-
-    if extend == 'both':
-        colors_i0 = 1
-        colors_i1 = -1
-        extra_colors = 2
-    elif extend == 'min':
-        colors_i0 = 1
-        extra_colors = 1
-    elif extend == 'max':
-        colors_i1 = -1
-        extra_colors = 1
-    elif extend == 'neither':
-        extra_colors = 0
-    else:
-        raise ValueError('Unexpected value for extend: {0!r}'.format(extend))
+    slice_map = {
+        'both': slice(1, -1),
+        'min': slice(1, None),
+        'max': slice(0, -1),
+        'neither': slice(0, None),
+    }
+    cbook._check_in_list(slice_map, extend=extend)
+    color_slice = slice_map[extend]
 
     n_data_colors = len(levels) - 1
-    n_expected_colors = n_data_colors + extra_colors
-    if len(colors) != n_expected_colors:
-        raise ValueError('With extend == {0!r} and n_levels == {1!r} expected'
-                         ' n_colors == {2!r}. Got {3!r}.'
-                         ''.format(extend, len(levels), n_expected_colors,
-                                   len(colors)))
-
-    cmap = ListedColormap(colors[colors_i0:colors_i1], N=n_data_colors)
+    n_expected = n_data_colors + color_slice.start - (color_slice.stop or 0)
```

Comment:
finding this a bit hard to follow ```(color_slice.stop or 0)```
-it always yields 0 or 1 and that gets subtracted?

Reply:
0 if stop is None (i.e. it goes all the way to the end), -(-1) if we drop the last element (so we need one more).